File: kotlin/src/main/kotlin/ai/muonium/muongit/Notes.kt
package ai.muonium.muongit

import java.io.File

/** Default notes reference */
const val DEFAULT_NOTES_REF = "refs/notes/commits"

/** A git note attached to an object */
data class Note(
    val noteOid: OID,
    val annotatedOid: OID,
    val message: String
)

/** Read a note for a specific object */
fun noteRead(gitDir: File, targetOid: OID, notesRef: String = DEFAULT_NOTES_REF): Note {
    val notesCommitOid = resolveReference(gitDir, notesRef)
    val (objType, data) = readLooseObject(gitDir, notesCommitOid)
    if (objType != ObjectType.COMMIT) throw MuonGitException.InvalidObject("notes ref not a commit")
    val commit = parseCommit(notesCommitOid, data)

    val noteOid = findNoteInTree(gitDir, commit.treeId, targetOid.hex)

    val (blobType, blobData) = readLooseObject(gitDir, noteOid)
    if (blobType != ObjectType.BLOB) throw MuonGitException.InvalidObject("note is not a blob")

    return Note(
        noteOid = noteOid,
        annotatedOid = targetOid,
        message = blobData.decodeToString()
    )
}

/** List all notes under a notes ref */
fun noteList(gitDir: File, notesRef: String = DEFAULT_NOTES_REF): List<Pair<OID, OID>> {
    val notesCommitOid = resolveReference(gitDir, notesRef)
    val (objType, data) = readLooseObject(gitDir, notesCommitOid)
    if (objType != ObjectType.COMMIT) throw MuonGitException.InvalidObject("notes ref not a commit")
    val commit = parseCommit(notesCommitOid, data)

    val notes = mutableListOf<Pair<OID, OID>>()
    collectNotesFromTree(gitDir, commit.treeId, "", notes)
    return notes
}

private fun findNoteInTree(gitDir: File, treeOid: OID, targetHex: String): OID {
    val (objType, data) = readLooseObject(gitDir, treeOid)
    if (objType != ObjectType.TREE) throw MuonGitException.InvalidObject("expected tree")
    val tree = parseTree(treeOid, data)

    if (targetHex.length >= 2) {
        val prefix = targetHex.substring(0, 2)
        val rest = targetHex.substring(2)

        for (entry in tree.entries) {
            if (entry.name == prefix && entry.mode == 0x4000) {
                return findNoteInTree(gitDir, entry.oid, rest)
            }
            if (entry.name == targetHex) return entry.oid
        }
    }

    for (entry in tree.entries) {
        if (entry.name == targetHex) return entry.oid
    }

    throw MuonGitException.NotFound("no note found for $targetHex")
}

private fun collectNotesFromTree(gitDir: File, treeOid: OID, prefix: String, notes: MutableList<Pair<OID, OID>>) {
    val (objType, data) = try {
        readLooseObject(gitDir, treeOid)
    } catch (_: Exception) { return }
    if (objType != ObjectType.TREE) return
    val tree = parseTree(treeOid, data)

    for (entry in tree.entries) {
        if (entry.mode == 0x4000) {
            collectNotesFromTree(gitDir, entry.oid, prefix + entry.name, notes)
        } else {
            val fullHex = prefix + entry.name
            if (fullHex.length == 40 && fullHex.all { it in '0'..'9' || it in 'a'..'f' }) {
                notes.add(entry.oid to OID(fullHex))
            }
        }
    }
}

File: swift/src/Notes.swift
/// MuonGit - Git notes: metadata annotations on commits
/// Parity: libgit2 src/libgit2/notes.c
import Foundation

/// Default notes reference
public let defaultNotesRef = "refs/notes/commits"

/// A git note attached to an object
public struct Note: Sendable {
    public let noteOid: OID
    public let annotatedOid: OID
    public let message: String
}

/// Read a note for a specific object
public func noteRead(gitDir: String, notesRef: String? = nil, targetOid: OID) throws -> Note {
    let ref = notesRef ?? defaultNotesRef

    let notesCommitOid = try resolveReference(gitDir: gitDir, name: ref)
    let (objType, data) = try readLooseObject(gitDir: gitDir, oid: notesCommitOid)
    guard objType == .commit else {
        throw MuonGitError.invalidObject("notes ref not a commit")
    }
    let commit = try parseCommit(oid: notesCommitOid, data: data)

    let noteOid = try findNoteInTree(gitDir: gitDir, treeOid: commit.treeId, targetHex: targetOid.hex)

    let (blobType, blobData) = try readLooseObject(gitDir: gitDir, oid: noteOid)
    guard blobType == .blob else {
        throw MuonGitError.invalidObject("note is not a blob")
    }
    guard let message = String(data: blobData, encoding: .utf8) else {
        throw MuonGitError.invalidObject("note is not valid UTF-8")
    }

    return Note(noteOid: noteOid, annotatedOid: targetOid, message: message)
}

/// List all notes under a notes ref
public func noteList(gitDir: String, notesRef: String? = nil) throws -> [(noteOid: OID, annotatedOid: OID)] {
    let ref = notesRef ?? defaultNotesRef

    let notesCommitOid = try resolveReference(gitDir: gitDir, name: ref)
    let (objType, data) = try readLooseObject(gitDir: gitDir, oid: notesCommitOid)
    guard objType == .commit else {
        throw MuonGitError.invalidObject("notes ref not a commit")
    }
    let commit = try parseCommit(oid: notesCommitOid, data: data)

    var notes: [(OID, OID)] = []
    try collectNotesFromTree(gitDir: gitDir, treeOid: commit.treeId, prefix: "", notes: &notes)
    return notes
}

// MARK: - Internal

private func findNoteInTree(gitDir: String, treeOid: OID, targetHex: String) throws -> OID {
    let (objType, data) = try readLooseObject(gitDir: gitDir, oid: treeOid)
    guard objType == .tree else {
        throw MuonGitError.invalidObject("expected tree")
    }
    let tree = try parseTree(oid: treeOid, data: data)

    if targetHex.count >= 2 {
        let prefix = String(targetHex.prefix(2))
        let rest = String(targetHex.dropFirst(2))

        for entry in tree.entries {
            if entry.name == prefix && entry.mode == 0o040000 {
                return try findNoteInTree(gitDir: gitDir, treeOid: entry.oid, targetHex: rest)
            }
        }
    }

    for entry in tree.entries {
        if entry.name == targetHex {
            return entry.oid
        }
    }

    throw MuonGitError.notFound("no note found for \(targetHex)")
}

private func collectNotesFromTree(gitDir: String, treeOid: OID, prefix: String, notes: inout [(OID, OID)]) throws {
    let (objType, data) = try readLooseObject(gitDir: gitDir, oid: treeOid)
    guard objType == .tree else { return }
    let tree = try parseTree(oid: treeOid, data: data)

    for entry in tree.entries {
        if entry.mode == 0o040000 {
            let newPrefix = "\(prefix)\(entry.name)"
            try collectNotesFromTree(gitDir: gitDir, treeOid: entry.oid, prefix: newPrefix, notes: &notes)
        } else {
            let fullHex = "\(prefix)\(entry.name)"
            if fullHex.count == 40 {
                let annotatedOid = OID(hex: fullHex)
                notes.append((entry.oid, annotatedOid))
            }
        }
    }
}
